### DISCUSSION TITLE
请问如何用wasi编译wasm项目？

### DISCUSSION THREAD

**Discussion by vkensou** [xmake-io/xmake#3085] [Q&A]
我研究了用emcc编译wasm的方法：

```
target("scripts")
    set_plat("wasm")
    set_toolchains("emcc")
    set_kind("binary")
    add_files("scripts/*.cpp")
    add_cleanfiles("scriptsout/*.*")
    add_cxflags("-O3")
    add_ldflags("-sERROR_ON_UNDEFINED_SYMBOLS=0")
    after_build(function(target)
        local dir = target:targetdir()
        local filename = target:basename() .. ".wasm"
        local filepath = path.join(dir, filename)
        local outdir = "scriptsout"
        os.mkdir(outdir)
        os.cp(filepath, outdir)
    end)
```

但没研究明白如何用wasi编译。应该怎么改呢？

**Comment by waruqi** [comment]
有专门的 wasi 工具链 https://github.com/xmake-io/xmake/pull/1737/files

**Reply by vkensou** [reply]
我尝试这么设置还是不行：
    set_plat("wasi")
    set_toolchains("wasi")

clang已经安装好，但编译的时候clang并没有带上 --target=wasm32 这个参数。不知道还需要设置什么。clang编译的时候会报错：fatal error LNK1561: 必须定义入口点。
我的这个wasm项目并不需要main函数。用emcc就一切正常。

**Reply by waruqi** [reply]
https://github.com/WebAssembly/wasi-sdk#use

> CC="${WASI_SDK_PATH}/bin/clang --sysroot=${WASI_SDK_PATH}/share/wasi-sysroot"

我没看到 wasi 需要 `--target=wasm32`

**Reply by vkensou** [reply]
> https://github.com/WebAssembly/wasi-sdk#use
> 
> > CC="${WASI_SDK_PATH}/bin/clang --sysroot=${WASI_SDK_PATH}/share/wasi-sysroot"
> 
> 我没看到 wasi 需要 `--target=wasm32`

不好意思，原来wasi里面也有clang。。。是我搞错了。但我现在把wasi添加到PATH里后，编译又报新的错误：
```
clang++ -o e:\myroom\test\wasmertest3\build\wasm32\x64\debug\scripts e:\myroom\test\wasmertest3\build\.objs\scripts\wasm32\x64\debug\scripts\hello.cpp.o --sysroot=nil\share\wasi-sysroot
wasm-ld: error: cannot open crt1.o: no such file or directory
wasm-ld: error: unable to find library -lc++
wasm-ld: error: unable to find library -lc++abi
wasm-ld: error: unable to find library -lc
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

现在报错找不到crt1.o
我已经把
D:\wasi-sdk\bin
D:\wasi-sdk
D:\wasi-sdk\share\wasi-sysroot
D:\wasi-sdk\share\wasi-sysroot\lib\wasm32-wasi
这几个路径都加入PATH里了。我看到xmake源码里是这么写的：
```
local sdkdir = toolchain:sdkdir()
local sysroot = path.join(sdkdir, "share", "wasi-sysroot")
```
但是wasi的toolchain文件里并没有调用set_sdkdir。我看编译命令是这样的：clang++ ****  --sysroot=nil\share\wasi-sysroot。不知道应该怎么设置才能让wasi找到这些lib？

**Reply by waruqi** [reply]
更新下 dev 再试试

```console
xmake f -p wasm --toolchain=wasi --sdk=/xxx/wasi-sdk -c
xmake
```

![image](https://user-images.githubusercontent.com/151335/203060668-fa3ef6d3-1394-4f6a-934d-0f7f7d641fc1.png)

**Reply by vkensou** [reply]
非常感谢。xmake 带上--sdk参数就可以编译了。但我还发现另外一些问题：
1.希望编译工具为wasi时，可以不用附带--sdk参数，由xmake自动检测。这个参数可以直接在xmake.lua里直接设置吗？
2.目前检测wasi toolchain的方法是不是和clang冲突了？毕竟clang和wasi-sdk里面都有clang.exe等文件；
3.platform为wasm时，targetfile的后缀都是 .html 。但是emcc会生成.wasm、.js和.html3个文件；而wasi生成的是.html，虽然这个文件实际上是.wasm。我看wasm的platform文件里是这么写的：set_formats("binary", "$(name).html")。
